### ISSUE TITLE
KeyError: 'pymarkups' on portable module

### ISSUE THREAD

**Issue by Tcll** [retext-project/pymarkups#13]
point in short I have ReText installed in a portable location and am using this module within this portable location...
but this module appears to need something extra in order to function properly:
![markups_KeyError](https://user-images.githubusercontent.com/773959/126914010-aad9e0d9-634b-4459-b95a-287cb2f2c98b.png)

if I could replace the `__getitem__("pymarkups")` call with a `get("pymarkups", default_value)` what would `default_value` be??

**Comment by Tcll** [comment]
update: so I've done some research with an updated Anaconda interpreter that can't load PyQt5 because the xcb plugin can't be found...

it would appear `entry_points()` tries to search a non-existant `<module-info>.dist-info/entry_points.txt` file using `importlib_metadata`.
This function returns a resulting tuple of `<class 'importlib_metadata.EntryPoint'>` objects as detected from `importlib_metadata`.

as I've installed a portable module relative to ReText, I lack the dist-info directory for `importlib_metadata` to search.

I guess a good `default_value` would be manually building a list of entry points relative to the extensions within the `markups` module.

I'll post the code/fix for this issue once I figure it out
for which you guys can create your own PR for. ;)

**Comment by Tcll** [comment]
well that was easier than I thought
for `markups.__init.py__`:
line 25: (not sure if you can just import `EntryPoint` on line 23, but you basically just need the class (probably from `types`))
`        from importlib_metadata import entry_points, EntryPoint` (I was only able to test this)
line 29:
```
        try: entrypoints = entry_points()["pymarkups"]
        except KeyError: entrypoints = (
            EntryPoint( name="markdown", value="markups.markdown:MarkdownMarkup", group="pymarkups" ),
            EntryPoint( name="restructuredtext", value="markups.restructuredtext:ReStructuredTextMarkup", group="pymarkups" ),
            EntryPoint( name="markups", value="markups.textile:TextileMarkup", group="pymarkups" )
        )
```

and it works, though it hits another error for me to fix:
<image coming soon, can no longer post without enabling malware>
it almost fully works though and displays a window for a frame.

also yes Wine is used because `scandir` (for `pathlib2`) is windows-only.

**Comment by mitya57** [comment]
You could just replace the whole content of `get_all_markups` function with:
```python
return builtin_markups
```

However I don’t want to support this use case. Most of other ReText dependencies *do* need to be properly installed anyway. If you don't want to install, please run `python3 setup.py egg_info`, this should be enough to get the entry points working.

Regarding the `horizontalAdvance` method, please make sure that both Qt and PyQt are at least 5.11 or newer, as documented in the README.

**Comment by Tcll** [comment]
thanks for the PyQt tip
but I'm actually backporting to PyQt4 since that doesn't give an xcb plugin issue (mentioned in the update post)
(all of my other interpreters (5 total) running PyQt5 have the xcb issue)

as for `builtin_markups`
I understand why you're relying on `importlib_metadata`
and while I'm not in favor of the system since it hampers portability
I'm also not entirely against it since other modules could potentially add entry points.

so I feel relying on them would be beneficial despite the builtin-markups being available

... though at the same time
taking what you said into account, I suppose I don't exactly need to compose entry points and can just `except KeyError: return builtin_markups` instead.
since that would support the edge case of portability, while prioritizing installation checks.

**Comment by mitya57** [comment]
> but I'm actually backporting to PyQt4 since that doesn't give an xcb plugin issue

Oh. Support for PyQt4 was removed seven (!) years ago in commit ace0e1991b43c97a9879361bca0b371020f46b82, so it will be hard to get it back.

If the `xcb` plugin cannot be found, try the following:

- Find where it is installed. You can try using `pip show -f PyQt5` to get the location and the list of installed files. The file name is `libqxcb.so`.
- Export `QT_PLUGIN_PATH` environment variable to point to the plugins path from the previous step.
- Check the `libqxcb.so` file with ldd and check if there are any missing references. If there are any, maybe you need to install some additional libraries.

**Comment by Tcll** [comment]
> `QT_PLUGIN_PATH`

oh hello there little abstract variable, apparently nobody on the internet seems to know exists...
... or searx is just an abhorrent search engine
I already have the path set (within launch.sh), but the export var is `LD_LIBRARY_PATH` as per common answers on various forums and SE questions.
big thanks there :+1:

although, I'm afraid you're a little late: XD
![ReText_window_first_display](https://user-images.githubusercontent.com/773959/127241266-10fadbca-a137-486f-aedb-17e19e5b82c1.png)
(the last issue was to replace `horizontalAdvance` with `width`)
^ all that was left from there was to just comment out a few things and uncategorize a few constants...
(eg: from `QTextEdit.LineWrapMode.fixedPixelWidth` to `QTextEdit.fixedPixelWidth`)
but for the most part, it's hardly incompatible :)

the only major things I've had to remove were print and PDF support, but I can probably add those back

also, as I'm running it with Wine (for `scandir`), I'm forced to use a python34 interpreter
(I could install newer, I'm just lazy, and don't see much reason to, as there's not much that's incompatible)
^ plus add to that, newer interpreters don't run in my windows test environment.
... so yeah I've just made ReText support more platforms :)

**Comment by mitya57** [comment]
Good job!

I still don't understand your configuration — if the host platform is Linux then why can't you run ReText natively? It doesn't need pathlib2 or scandir…

Also maybe using some older ReText version as a base will be easier, you won't need to revert scoped enums and other things.

**Comment by Tcll** [comment]
> I still don't understand your configuration — if the host platform is Linux then why can't you run ReText natively? It doesn't need pathlib2 or scandir…

it's `importlib_metadata` that uses `pathlib2`, which uses `scandir`

I've actually updated things with another try/except on `importlib_metadata` and also returned `builtin_markups` if that check failed. (if you have the module, it'll be used)
so I've now removed those 3 modules, which should allow me to run it with a native interpreter :)

as for the configuration itself, I'm treating ReText as a portable program, with all required dependencies included in it's CWD.
some of which I've had to copy from other interpreters like `enum` and `typing` and modify for py34 just to get the window displayed.
![Screenshot_2021-07-28_09-18-56](https://user-images.githubusercontent.com/773959/127329885-625c2f65-9b0e-4345-9aa2-829bbb3aa356.png)
^ everything relative to the program is included here.
(this is a practice I've taken up a long while back since deps constantly prove to be unreliable and break your program on updates)
^ though a better practice I rely on is to just not use deps if I can avoid it.
(for example, `enum` can probably be removed since it's only used for 1 class)

> Also maybe using some older ReText version as a base will be easier, you won't need to revert scoped enums and other things.

I may look back on an older version if I need to :)
but 18 modules averaging 6KiB is tiny (moderately sized: `config`: 11KiB, `editor`: 22KiB, `tab`: 19KiB, `window`: 49KiB)
so this is a decently easy overhaul ;)

after stripping down PyOpenGL into my own efficient cbin.GL (ctypes itself is also inefficient), yeah, this is nothing ;)
plus I have PyQt4 experience to back :+1: 
(I don't like Qt, but I understand why you'd use it. though FLTK is better if you're willing to make a port)
